Library/Homebrew/unpack_strategy/directory.rb

```diff
@@ -20,29 +20,34 @@ def self.can_extract?(path)
 
     sig { override.params(unpack_dir: Pathname, basename: Pathname, verbose: T::Boolean).void }
     def extract_to_dir(unpack_dir, basename:, verbose:)
-      path_children = path.children
-      return if path_children.empty?
-
-      existing = unpack_dir.children
-
-      # We run a few cp attempts in the following order:
-      #
-      # 1. Start with `-al` to create hardlinks rather than copying files if the source and
-      #    target are on the same filesystem. On macOS, this is the only cp option that can
-      #    preserve hardlinks but it is only available since macOS 12.3 (file_cmds-353.100.22).
-      # 2. Try `-a` as GNU `cp -a` preserves hardlinks. macOS `cp -a` is identical to `cp -pR`.
-      # 3. Fall back on `-pR` to handle the case where GNU `cp -a` failed. This may happen if
-      #    installing into a filesystem that doesn't support hardlinks like an exFAT USB drive.
-      cp_arg_attempts = ["-a", "-pR"]
-      cp_arg_attempts.unshift("-al") if path.stat.dev == unpack_dir.stat.dev
-
-      cp_arg_attempts.each do |arg|
-        args = [arg, *path_children, unpack_dir]
-        must_succeed = print_stderr = (arg == cp_arg_attempts.last)
-        result = system_command("cp", args:, verbose:, must_succeed:, print_stderr:)
-        break if result.success?
-
-        FileUtils.rm_r(unpack_dir.children - existing)
+      path.find(ignore_error: false) do |src|
+        next if src == path
+
+        dst = unpack_dir/src.relative_path_from(path)
+
+        if dst.directory? && !dst.symlink?
+          # Output same error as `cp` when trying to copy over an existing directory
+          raise "#{dst}: Is a directory" if !src.directory? || src.symlink?
+
+          begin
+            # Fix group caused by unpacking into tmp dir. Try to copy valid group over
+            FileUtils.chown(nil, src.lstat.gid == "wheel" ? dst.parent.lstat.gid : src_gid, dst)
```

Comment:
Would be nice to use an `if`/`else` here, bit hard to read as-is

Reply:
Will be trying to fix this earlier in #18548.